Warn user if regular expression is used when match_pattern defaults to simple in dynamic_template mappings

quux00 opened the pull request:
In a dynamic_templates setting, when match_pattern is not defined by the user and thus defaults to "simple" and when the pattern passed into match, unmatch, path_match or path_unmatch appears to be a regular expression, 
a warning is added to the HTTP response header. The warning will not be set if the user explicitly sets match_pattern=simple, since ES allows most regular expression meta-characters in doc field names.

Closes #95634

quux00 committed:
Warn user if regular expression is used when match_pattern defaults to simple in dynamic_template mappings

In a dynamic_templates setting, if match_phrase=simple (explicitly or by default) and the pattern
passed into match, unmatch, path_match or path_unmatch appears to be a regular expression instead
a warning is added to the HTTP response header.

Closes #95634

quux00 committed:
Removed isESFieldRegexPattern from Regex.java

quux00 committed:
Added MatchType.DEFAULT

This avoids losing information on multiple parses as to whether the user supplied
the SIMPLE setting or whether just defaulted to it.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

quux00 commented:
Tested this locally with curl and warning appears as expected:

```
└─ $ ▶ cat > dyntemplate_with_regex.json
{
  "mappings": {
    "dynamic_templates": [
      {
        "mytemplate": {
          "match": ["^12.*", "foo*"],
          "mapping": {
            "type": "keyword"
          }
        }
      }
    ]
  }
}

└─ $ ▶ curlj -X PUT localhost:9200/myindex -d@dyntemplate_with_regex.json
HTTP/1.1 200 OK
X-elastic-product: Elasticsearch
Warning: 299 Elasticsearch-8.9.0-SNAPSHOT-0767ebd34cb70cea8676563f774ba22038937a80 "Pattern [^12.*] appears to be a regular expression, not a simple wildcard pattern. To remove this warning, set [match_pattern] in the dynamic template."
content-type: application/json
content-length: 66

{"acknowledged":true,"shards_acknowledged":true,"index":"myindex"}
```

dnhatn commented:
@elasticsearchmachine update branch

dnhatn commented:
update branch

dnhatn commented:
@elasticmachine update branch

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -60,6 +146,12 @@ public static MatchType fromString(String value) {
+         * Whether the {@code pattern} is a valid for the MatchType.
+         * @throws IllegalArgumentException if the pattern is not valid
+         */
+        public abstract void isValidPattern(String pattern);
```
I think I'd call this `validate` - `isXXX` implies a boolean return value to me.

quux00 replied:
Yeah, good call. The original version I did returned boolean, but I need the error message on failure. I'll change it.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -255,19 +347,26 @@ private static void addEntriesToPatternList(List<String> matchList, String propN
-                    ),
-                    e
-                );
+                if (matchType == MatchType.DEFAULT) {
```
Can we move this logic into `isValidPattern` directly, rather than throwing and catching exceptions?

romseygeek replied:
So the reason I think this fits better in `validate` is that there are two branches inside this catch clause, and each only applies to one match type.  There's no shared code here.  In other words, these are bits of MatchType-specific logic, which to me means that they should live within the different MatchType implementations.

quux00 replied:
Sounds good. I'll move it.

quux00 replied:
This change required more code changes that I expected. Key changes:

1) I added templateName to the `MatchType.validate` params list, because the error message (that was already in place before I started this series of tickets) requires the template name in the error message.

2) Testing `MatchType.validate` with only side effects and no return value became problematic, so I changed it to return a boolean (useful mostly for unit tests) and thus changed the name back to `isValidPattern`.

quux00 replied:
Actually, nevermind the second comment above. After consideration I found that testing without a boolean return value is straightforward. Pushed new commit.

romseygeek commented on server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java:
```diff
@@ -2252,4 +2373,137 @@ public void testInvalidMatchWithArrayOfFieldNamesUsingNonStringEntries() throws
+    }
+
+    public void testIsValidPatternForSimpleMatchType() {
+        record MatchPattern(String pattern) {}
```
I'm not sure this adds anything over just using `String` here?

quux00 replied:
Right. I'll change that.

romseygeek reviewed (changes requested):
Thanks @quux00!  I left a few comments.

quux00 committed:
PR changes: renamed MatchType.isValidPattern to validate

quux00 committed:
Second PR round: changes to DynamicTempalte.MatchType.validate

Renamed Match.Type.validate to isValidPattern and
moved logic of header warnings and user exceptions there and out of DynamicTemplate.parse.

This required adding a boolean return value to the method for testing purposes.
And requires passing in the templateName for error/warning messages.

quux00 committed:
Remove unused method, missed in the last PR commit

quux00 committed:
Realized I can test without returning boolean from isValidPattern, so reverted to void validate signature

javanna commented:
I have a general question around the overall idea of emitting a warning: is the plan to one day remove the warning and make it a hard error? What type of warning are we emitting in the proposed change, is that a deprecation warning?

romseygeek reviewed (approved):
LGTM, thanks for all the iterations @quux00!

quux00 commented:
> The idea here is that if you don't explicitly set `match_type`, then we use simple globbing to match field paths. But it's fairly easy to overlook this and try and use a regex as a pattern without setting `match_type`, which can fail in unexpected ways - and importantly, not fail until quite a long time after you've set things up. So the idea of this warning is to detect cases where somebody hasn't explicitly set the `match_type`, but looks like they're using a regex. It would never become a hard error, because it's only a heuristic - we think you're using a regex here but you haven't set it up properly. If it is in fact correctly configured (ie you really do want to match fields that have square brackets in them), then you can make the warning go away by setting `match_type=simple`.

Good write-up. I'll adjust the commit message to reflect this context.
